[Runtime] Fix strong and unowned refcount overflow on 32-bit.

mikeash opened the pull request:
Fix overflow detection on unowned refcounts so that we create a side table when incrementing from 126. Implement strong refcount overflow to the side table.

The unowned refcount is never supposed to be 127, because that (sometimes) represents the immortal refcount. We attempt to detect that by checking newValue == Offsets::UnownedRefCountMask, but the mask is shifted so that condition is never true. We managed to hit the side table case when incrementing from 127, because it looks like the immortal case. But that broke when we fixed immortal side table initialization in b41079a8f54ae2d61c68cdda46c74232084af020. With that change, we now create an immortal side table when overflowing the unowned refcount, then try to increment the unowned refcount in that immortal side table, which traps.

rdar://123788910

mikeash committed:
[Runtime] Fix strong and unowned refcount overflow on 32-bit.

Fix overflow detection on unowned refcounts so that we create a side table when incrementing from 126. Implement strong refcount overflow to the side table.

The unowned refcount is never supposed to be 127, because that (sometimes) represents the immortal refcount. We attempt to detect that by checking newValue == Offsets::UnownedRefCountMask, but the mask is shifted so that condition is never true. We managed to hit the side table case when incrementing from 127, because it looks like the immortal case. But that broke when we fixed immortal side table initialization in b41079a8f54ae2d61c68cdda46c74232084af020. With that change, we now create an immortal side table when overflowing the unowned refcount, then try to increment the unowned refcount in that immortal side table, which traps.

rdar://123788910

Catfish-Man reviewed (approved):
I can't in good faith say "looks good to me" about anything in these files, but this looks correct, thank you for doing it

al45tair reviewed (approved):
Looks good.

al45tair commented:
@swift-ci Please smoke test

mikeash commented on test/Interpreter/refcount_overflow.swift:
```diff
@@ -0,0 +1,163 @@
+RefcountOverflowTests.test("retain") {
+  do {
+    let obj = C()
+    _ = wrapper_swift_retain_n(unsafeBitCast(obj, to: UnsafeMutableRawPointer.self), maxRC - 2)
```
I thought this was the right value, but it turns out it was just ARC inserting surplus retains when optimizations are off. I've fixed the tests to immediately convert the object to a raw pointer and do all work with that to avoid ARC causing trouble.

mikeash commented on stdlib/public/SwiftShims/swift/shims/RefCount.h:
```diff
@@ -397,7 +397,7 @@ class RefCountBitsT {
     auto newValue = getUnownedRefCount();
     return newValue != oldValue + inc ||
-      newValue == Offsets::UnownedRefCountMask;
+      newValue == Offsets::UnownedRefCountMask >> Offsets::UnownedRefCountShift;
```
I think we'd do better using bitfields instead of a bunch of constants and shifts, then we wouldn't need masks and could just deal in a plain "max value." But that would be a major overhaul. Short of that, I can't think of anything except for better test coverage. Hopefully we'll be in a bit better shape in that respect now.

mikeash replied:
It might be acceptable with a union of multiple different bitfield layouts. We'd need one with `strong: 22; useSlowRC: 1` and one with `strongAndUseSlowRC: 23`, for example. That might end up just being worse.

al45tair replied:
That'd be UB these days though; you'd have to add some `memcpy()`s, which hopefully would get optimised away, but still.

We could also have done a "fake bitfield" thing with operator overloads and special types that actually do the relevant shifting and masking under the covers. I didn't think that was much better either :-)

mikeash commented:
Updated with improved tests that aren't dependent on ARC optimizations, cover the case where we overflow the unowned refcount by a moderately large value (I thought that might be broken when looking at the code; it turns out to be fine, but now we have ongoing proof), and removed the now-redundant and less comprehensive `unowned_overflow.swift' test.

mikeash commented:
@swift-ci please test

mikeash commented:
@swift-ci please test

mikeash commented:
@swift-ci please smoke test

mikeash commented:
Mac tests hit a network error cloning the source, trying again.

mikeash commented:
@swift-ci please smoke test
